public/client.js

```diff
@@ -0,0 +1,262 @@
+	if(type == "input")
+		log("name", name, "port", port, "state", state);
+
+}
+
+function listInputs(inputs){
+	var input = inputs.value;
+		log("Input port : [ type:'" + input.type + "' id: '" + input.id +
+				"' manufacturer: '" + input.manufacturer + "' name: '" + input.name +
+				"' version: '" + input.version + "']");
+}
+
+function noteOn(midiNote, velocity){
+	player(midiNote, velocity);
+}
+
+function noteOff(midiNote, velocity){
+	player(midiNote, velocity);
+}
+
+function player(note, velocity){
+	var sample = sampleMap['key'+note];
+	if(sample){
+		if(type == (0x80 & 0xf0) || velocity === 0){
+			btn[sample - 1].classList.remove('active');
+			return;
+		}
+		btn[sample - 1].classList.add('active');
+		btn[sample - 1].play(velocity);
+	}
+}
+
+function onMIDIFailure(e){
+	log("No access to MIDI devices or your browser doesn't support WebMIDI API. Please use WebMIDIAPIShim " + e);
+}
+
+// MIDI utility functions
+function showMIDIPorts(midiAccess){
+	var inputs = midiAccess.inputs,
+			outputs = midiAccess.outputs,
+			html;
+	html = '<h4>MIDI Inputs:</h4><div class="info">';
+	inputs.forEach(function(port){
+		html += '<p>' + port.name + '<p>';
+		html += '<p class="small">connection: ' + port.connection + '</p>';
+		html += '<p class="small">state: ' + port.state + '</p>';
+		html += '<p class="small">manufacturer: ' + port.manufacturer + '</p>';
+		if(port.version){
+			html += '<p class="small">version: ' + port.version + '</p>';
+		}
+	});
+	deviceInfoInputs.innerHTML = html + '</div>';
+
+	html = '<h4>MIDI Outputs:</h4><div class="info">';
+	outputs.forEach(function(port){
+		html += '<p>' + port.name + '<br>';
+		html += '<p class="small">manufacturer: ' + port.manufacturer + '</p>';
+		if(port.version){
+			html += '<p class="small">version: ' + port.version + '</p>';
+		}
+	});
+	deviceInfoOutputs.innerHTML = html + '</div>';
+}
+
+// audio functions
+function loadAudio(object, url){
+	var request = new XMLHttpRequest();
+	request.open('GET', url, true);
+	request.responseType = 'arraybuffer';
+	request.onload = function(){
+		context.decodeAudioData(request.response, function(buffer){
+			object.buffer = buffer;
+		});
+	};
+	request.send();
+}
+
+function addAudioProperties(object){
+	object.name = object.id;
+	object.source = object.dataset.sound;
+	loadAudio(object, object.source);
+	object.play = function(volume){
+		var s = context.createBufferSource();
+		var g = context.createGain();
+		var v;
+		s.buffer = object.buffer;
+		s.playbackRate.value = randomRange(0.5, 2);
+		if(volume){
+			v = rangeMap(volume, 1, 127, 0.2, 2);
+			s.connect(g);
+			g.gain.value = v * v;
+			g.connect(context.destination);
+		}
+		else{
+			s.connect(context.destination);
+		}
+
+		s.start();
+		object.s = s;
+	};
+}
+
+// utility functions
+function randomRange(min, max){
+	return Math.random() * (max + min) + min;
+}
+
+function rangeMap(x, a1, a2, b1, b2){
```

Comment:
Looks pretty interesting, my only question is for the rangeMap function could you have better named variables so that it's easier to tell what is being calculated there?